### PR TITLE
Add email regexp 231

### PR DESCRIPTION
**Pull request by inhundreds**
## Summary of change

- Added a regexp to validate User email address
- Added a test for invalid email address
- Fixed the valid user test adding a valid mail to the user used for the test

## Pull request checklist

Please ensure you have done the following:

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request

- [x] Created a branch that has a descriptive name (what your branch is for a in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

If in doubt, get in touch with us via our [Gitter room](https://gitter.im/book-project-community/community)

### CODE REVIEW

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java [hunk 1]
```diff
@@ -62,6 +62,12 @@
 
     // Note: this is allowed to be null if a user signs up without an email
     @NotEmpty
+    // For the RegExp see https://owasp.org/www-community/OWASP_Validation_Regex_Repository
+    @Pattern(
+            regexp = "^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$",
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/User.java:67]: You seem to be using the wrong number of backslashes at certain points. Could you replace the regex with the following (taken from OWASP):
`[a-zA-Z0-9_+&*-]+(?:\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,7}$]]`

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java [hunk 2]
```diff
@@ -27,7 +27,7 @@ void beforeEachSetup() {
     @Test
     void testValidUserSaved() {
         // given
-        User user = userWithoutPassword().password("passwordP1&")
+        User user = userWithoutPassword().password("passwordP1&").email("abc@def.com")
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:30]: Could you drop the `.email` method invocation down one line (aligned with `.password`)?

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java [hunk 3]
```diff
@@ -115,4 +122,16 @@ void testPasswordWithNoSpecialCharacterIsInvalid() {
         // then
         Assertions.assertEquals(initialNumberOfUsers, userRepository.count());
     }
+
+    @Test
+    void testUserWithInvalidMail() {
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:127]: Could you rename this method to be more explicit about what is being tested? While we know a user with an invalid email address is being tested, we do not know why exactly the email is invalid? 

This is important because other invalid email tests are required.

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java [hunk 4]
```diff
@@ -115,4 +122,16 @@ void testPasswordWithNoSpecialCharacterIsInvalid() {
...
+        // given
+        User user = userWithInvalidMail().build();
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:129]: Could you delete the `userWithInvalidMail()` method and instead create a new method called `createUserWithNoEmail()` (or something similar)?

Then, you can use that method and specify the email in this method.

Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:129]: As alluded to before, it's great that we're testing that email addresses without a domain are not accepted. However, we have not yet fully tested the regex. For example, other tests could include email addresses that do not include the `@` sign, email addresses without a top-level domain, email addresses with nothing before the `@` sign and so on. 

In short, we want to fully test all of the ways that someone may try to get around the regex (rather than simply looking at the regex and figuring out the edge cases). This way, we can have a higher level of surety that our email regex works as desired.

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java (hunk 1 above)
Comment (inhundreds) [reply src/main/java/com/karankumar/bookproject/backend/entity/User.java:67]: Thanks for your comments! One backslash is needed to escape the dot and the other is needed to backslash the first. Take a look here: https://stackoverflow.com/a/3675027/7885311.

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java (hunk 4 above)
Comment (inhundreds) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:129]: Ok, I will extract the regex as a constant in  a utility class and I will add some tests (in a new test file) to improve our confidence level about the regex. Do you think is it ok?

Thanks,
Diego

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java (hunk 1 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/entity/User.java:67]: My bad. Ignore me :)

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java (hunk 4 above)
Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:130]: I think it's best to keep the regex where it is for now and to not create a new test file for the time being.

Comment (inhundreds) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java:130]: Ok, thanks. I will work on it in the next days.

**Comment by knjk04** [comment]
In case you weren't aware, you won't show up as a contributor (see https://docs.github.com/en/github/setting-up-and-managing-your-github-profile/why-are-my-contributions-not-showing-up-on-my-profile)

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java (hunk 4 above)
Comment (inhundreds) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java]: I added tests for some invalid cases but they will never be exhaustive. Let me know if some other case comes up to your mind.

Thanks,
Diego

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java (hunk 2 above)
Comment (inhundreds) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java]: Yes, I fixed it

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java (hunk 3 above)
Comment (inhundreds) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java]: Yes, I fixed and added other tests.

**Comment by knjk04** [comment]
Please could you resolve the merge conflicts?

File: src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java (hunk 4 above)
Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/entity/UserTest.java]: The additions are good. Thanks a lot!

**Comment by inhundreds** [comment]
Thanks to you! I hope to contribute again soon!
